File: backend/app/schemas/reports.py
"""Reports schemas for request/response validation."""

from datetime import date
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


class ProjectProgressReport(BaseModel):
    """Project progress report."""

    project_id: str
    project_name: str
    project_key: str
    status: str
    total_tasks: int
    completed_tasks: int
    in_progress_tasks: int
    todo_tasks: int
    overdue_tasks: int
    progress_percentage: float
    total_estimated_hours: float
    total_actual_hours: float
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class UserTimeReport(BaseModel):
    """Time logged per user."""

    user_id: str
    full_name: str
    email: str
    avatar_url: Optional[str] = None
    total_minutes: int
    total_hours: float
    task_count: int


class TimeReportResponse(BaseModel):
    """Monthly time report response."""

    month: int
    year: int
    total_hours: float
    user_reports: List[UserTimeReport]


class OverviewReportResponse(BaseModel):
    """Full overview report across all projects."""

    total_projects: int
    active_projects: int
    completed_projects: int
    total_tasks: int
    completed_tasks: int
    overdue_tasks: int
    overall_progress: float
    projects: List[ProjectProgressReport]
    top_contributors: List[UserTimeReport]

File: app/src/api/reports.ts
// Reports API client

import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';

function authHeaders() {
  return { Authorization: `Bearer ${localStorage.getItem('token')}` };
}

export interface ProjectProgressReport {
  project_id: string;
  project_name: string;
  project_key: string;
  status: string;
  total_tasks: number;
  completed_tasks: number;
  in_progress_tasks: number;
  todo_tasks: number;
  overdue_tasks: number;
  progress_percentage: number;
  total_estimated_hours: number;
  total_actual_hours: number;
  start_date?: string;
  end_date?: string;
}

export interface UserTimeReport {
  user_id: string;
  full_name: string;
  email: string;
  avatar_url?: string;
  total_minutes: number;
  total_hours: number;
  task_count: number;
}

export interface TimeReportData {
  month: number;
  year: number;
  total_hours: number;
  user_reports: UserTimeReport[];
}

export interface OverviewReportData {
  total_projects: number;
  active_projects: number;
  completed_projects: number;
  total_tasks: number;
  completed_tasks: number;
  overdue_tasks: number;
  overall_progress: number;
  projects: ProjectProgressReport[];
  top_contributors: UserTimeReport[];
}

export async function fetchProjectProgress(projectId: string): Promise<ProjectProgressReport> {
  const response = await axios.get(`${API_URL}/reports/project-progress/${projectId}`, {
    headers: authHeaders(),
  });
  return response.data;
}

export async function fetchTimeSummary(): Promise<TimeReportData> {
  const response = await axios.get(`${API_URL}/reports/time-summary`, {
    headers: authHeaders(),
  });
  return response.data;
}

export async function fetchOverviewReport(): Promise<OverviewReportData> {
  const response = await axios.get(`${API_URL}/reports/overview`, {
    headers: authHeaders(),
  });
  return response.data;
}
